conwayste/src/ui/layer.rs

```diff
@@ -31,137 +31,251 @@ use super::{
-    pub fn new(widget_id: WidgetID) -> Self {
-        Layer {
-            id: widget_id,
-            widgets: vec![],
+/// A container of one or more widgets or panes ordered by virtual layers.
+impl Layering {
+    pub fn new() -> Self {
+        Layering {
             with_transparency: false,
-            focused_widget: None,
+            widget_list: vec![],
+            focused_ids: vec![None],
+            id_cache: vec![]
         }
     }
 
-    /// Add a widget to the layer
-    pub fn add(&mut self, widget: Box<dyn Widget>) {
-        self.widgets.push(widget);
+    /// Returns true if an entry with the provided WidgetID exists.
+    fn check_for_entry(&self, widget_id: WidgetID) -> bool {
+        self.id_cache
+            .iter()
+            .find(|&&id| id == widget_id)
+            .is_some()
     }
 
-    /// Layer passes `exit_focus` request forward to its elements
-    pub fn exit_focus(&mut self) {
-        if let Some(other_id) = self.focused_widget {
-            if let Some(other_tf) = TextField::widget_from_id(self, other_id)
-            {
-                other_tf.exit_focus();
+    /// Returns an optional pair of indices if the widget-id is found in Pane beloning to
+    /// the layering.
+    fn search_panes_for_widget_id(&self, widget_id: WidgetID) -> Option<(usize, usize)> {
+        // First check to see if it belongs to any pane
+        for (i, w) in self.widget_list.iter().enumerate() {
+            if let Some(pane) = downcast_widget!(w, Pane) {
+                for (j, w2) in pane.widgets.iter().enumerate() {
+                    if w2.id() == widget_id {
+                        return Some((i, j));
+                    }
+                }
             }
         }
+
+        None
     }
 
-    /// Layer passes `enter_focus` request forward to its elements
-    pub fn enter_focus(&mut self, id: WidgetID) -> UIResult<()> {
-        self.exit_focus();
+    /// Caches the widget-id's residing in this layering. Will include widgets belonging to a one
+    /// level deep pane.
+    fn rebuild_id_cache(&mut self) {
+        let mut local_id_cache = vec![];
+        for widget in self.widget_list.iter() {
+            local_id_cache.push(widget.id());
+
+            if let Some(pane) = downcast_widget!(widget, Pane) {
+                local_id_cache.extend(pane.get_widget_ids());
+            }
+        }
+        self.id_cache = local_id_cache;
+    }
 
-        if let Some(tf) = TextField::widget_from_id(self, id) {
-            tf.enter_focus();
-            self.focused_widget = Some(id);
-            return Ok(());
+    /// Retreives a mutable reference to a widget. This will search one Pane-level deep
+    /// for the provided widget-id.
+    ///
+    /// # Error
+    /// A WidgetNotFound error will be returned if the widget-id is not found.
+    /// does not exist in the internal list of widgets.
+    pub fn get_widget_mut(&mut self, widget_id: WidgetID) -> UIResult<&mut Box<dyn Widget>> {
```

Comment:
Thoughts on using something like this?
```rust
type BoxedWidget = Box<dyn Widget>;
```
and then we can do
```rust
 pub fn get_widget_mut(&mut self, widget_id: WidgetID) -> UIResult<&mut BoxedWidget>
```

Reply:
I was thinking something like `&mut dyn Widget` so there is only one level of indirection.

See for example:
```rust
trait T {
    fn some_method(&mut self) -> i64;
}

#[derive(Default)]
struct S {
    pub x: i64,
}

impl T for S {
    fn some_method(&mut self) -> i64 {
        self.x += 1;
        self.x
    }
}

fn gets_a_t(tref: &mut dyn T) {
    tref.some_method();
}

fn main() {
    let mut box_t: Box<dyn T> = Box::new(S::default());
    gets_a_t(&mut *box_t);
}
```

Reply:
That might be a bit tricky given that it's being returned, and we don't know the size of them (hence the `Box` and not `Sized`) for all `Widgets`. I don't see a clear path without trying out some things at least. 

Could we address this in a separate issue since this PR is nearly done?

Reply:
OK